administration/src/util/result.ts

```diff
@@ -0,0 +1,96 @@
+type ResultError = { type: string }
+
+abstract class _Result<T, E extends ResultError> {
+    protected abstract _chain<X, U extends ResultError>(
+        ok: (value: T) => Result<X, U>,
+        err: (error: E) => Result<X, U>
+    ): Result<X, U>;
+
+    unwrap(): T;
+    unwrap<U>(ok: (value: T) => U): U;
+    unwrap<U, V>(ok: (value: T) => U, err: (error: E) => V): U | V;
+    unwrap<U>(ok: (value: T) => U, err: (error: E) => U): U;
+    unwrap(ok?: (value: T) => unknown, err?: (error: E) => unknown): unknown {
+        const r = this._chain(
+            value => Ok(ok ? ok(value) : value),
+            error => (err ? Ok(err(error)) : Err(error))
+        );
+        if (r.isErr) {
+            throw r.error;
+        }
+        return r.value;
+    }
+
+    map<U>(ok: (value: T) => U): Result<U, E>;
+    map<U, F extends ResultError>(
+        ok: (value: T) => U,
+        err: (error: E) => F
+    ): Result<U, F>;
+    map(ok: (value: T) => unknown, err?: (error: E) => ResultError): Result<unknown> {
+        return this._chain(
+            value => Ok(ok(value)),
+            error => Err(err ? err(error) : error)
+        );
+    }
+
+    chain<X>(ok: (value: T) => Result<X, E>): Result<X, E>;
+    chain<X>(
+        ok: (value: T) => Result<X, E>,
+        err: (error: E) => Result<X, E>
+    ): Result<X, E>;
+    chain<X, U extends ResultError>(
+        ok: (value: T) => Result<X, U>,
+        err: (error: E) => Result<X, U>
+    ): Result<X, U>;
+    chain(
+        ok: (value: T) => Result<unknown>,
+        err?: (error: E) => Result<unknown>
+    ): Result<unknown> {
+        return this._chain(ok, err || (error => Err(error)));
+    }
+}
+
+class _Ok<T, E extends ResultError> extends _Result<T, E> {
+    readonly isOk = true;
+    readonly isErr = false;
+    constructor(readonly value: T) {
+        super();
+    }
+
+    protected _chain<X, U extends ResultError>(
+        ok: (value: T) => Result<X, U>,
+        _err: (error: E) => Result<X, U>
+    ): Result<X, U> {
+        return ok(this.value);
+    }
+}
+
+class _Err<T, E extends ResultError> extends _Result<T, E> {
+    readonly isOk = false;
+    readonly isErr = true;
+    constructor(readonly error: E) {
+        super();
+    }
+
+    protected _chain<X, U extends ResultError>(
+        _ok: (value: T) => Result<X, U>,
+        err: (error: E) => Result<X, U>
+    ): Result<X, U> {
+        return err(this.error);
+    }
+}
+
+export function Ok<T, E extends ResultError>(value: T): Result<T, E> {
+    return new _Ok(value);
+}
+
+export function Err<E extends ResultError, T = never>(error: E): Result<T, E> {
+    return new _Err<T, E>(error);
+}
+
+/**
+ * This type was adopted from https://github.com/badrap/result
+ */
+export type Result<T, E extends ResultError = ResultError> =
+    | _Ok<T, E>
+    | _Err<T, E>;
```

Comment:
Is there a reason to copy&paste the code instead of using the npm version?
Maybe you want to document the changes you made to the original code here.

Reply:
Yes, it would add a third party dependency. The code we are using now is very simple in my opinion. I would even say it is pretty much bug-less and feature-complete.

I think there is no need to document changes. Let's just pretend I wrote all myself.